server/bleep/src/webserver/aaa.rs

```diff
@@ -0,0 +1,341 @@
+    responses(
+        (status = 200, description = "Execute query successfully", body = Response),
+        (status = 400, description = "Bad request", body = EndpointError),
+        (status = 500, description = "Server error", body = EndpointError),
+    ),
+)]
+pub(super) async fn login(
+    Extension(app): Extension<Application>,
+    Extension(auth_layer): Extension<Arc<AuthLayer>>,
+) -> impl IntoResponse {
+    let state = rand::thread_rng()
+        .sample_iter(Alphanumeric)
+        .take(STATE_LEN)
+        .map(|c| c as char)
+        .collect::<String>();
+
+    let client_id = app
+        .config
+        .github_client_id
+        .as_ref()
+        .expect("github client id must be provided")
+        .expose_secret();
+
+    let redirect_uri = format!(
+        "https://{}/auth/login/complete",
+        app.config
+            .instance_domain
+            .as_ref()
+            .expect("instance domain must be provided")
+    );
+
+    let github_oauth_url = &format!(
+        "https://github.com/login/oauth/authorize\
+         ?client_id={client_id}\
+         &state={state}\
+         &allow_signup=false\
+         &redirect_uri={redirect_uri}",
+    );
+
+    auth_layer.initialized_login.insert(state, Instant::now());
+    auth_layer.clean_old_states();
+
+    serde_json::json!({ "oauth_url": github_oauth_url }).to_string()
+}
+
+#[derive(Deserialize)]
+pub(super) struct AuthorizedParams {
+    state: State,
+    code: String,
+}
+
+/// Complete the login flow.
+///
+/// Takes the `state` that has been established previously so we don't leak the actual keys.
+#[utoipa::path(post, path = "/auth/login/complete",
+    responses(
+        (status = 200, description = "Execute query successfully", body = Response),
+        (status = 400, description = "Bad request", body = EndpointError),
+        (status = 500, description = "Server error", body = EndpointError),
+    ),
+)]
+pub(super) async fn authorized(
+    Extension(app): Extension<Application>,
+    Extension(auth_layer): Extension<Arc<AuthLayer>>,
+    Query(params): Query<AuthorizedParams>,
+    jar: PrivateCookieJar,
+) -> impl IntoResponse {
+    let AuthorizedParams { state, code } = params;
+
+    auth_layer
+        .initialized_login
+        .remove(&state)
+        .expect("invalid state key");
+
+    let (client_id, client_secret) = app
+        .config
+        .github_client_id_and_secret()
+        .expect("github client id and secret must be provided");
+
+    let gh_token = auth_layer
+        .client
+        .post(format!(
+            "https://github.com/login/oauth/access_token\
+            ?client_id={client_id}\
+            &client_secret={client_secret}\
+            &code={code}"
+        ))
+        .header("Accept", "application/json")
+        .send()
+        .await
+        .unwrap()
+        .json()
+        .await
+        .unwrap();
+
+    (
+        jar.add(AuthCookie::new(gh_token).to_cookie()),
+        Redirect::to("/"),
```

Comment:
can you also remove this redirect and instead create an endpoint that will just return true if the user is logged in, so that on the FE we know if it's time to show the next step?

Reply:
This is the redirect target for GitHub's login page; after logging in we hit this route, which ultimately redirects to `/`. It's not meant to be called directly as part of an API call.